src/extension.ts

```diff
@@ -163,10 +163,11 @@ export function activate(context: vs.ExtensionContext) {
 		activeFileFilters.push(HTML_MODE);
 	}
 
+	const triggerCharacters = ".: =(${'\"abcdefghijklmnopqrstuvwxyz".split("");
```

Comment:
I don't like the idea of adding the whole alphabet - what was this to try and do?

My concern is that if a user is in the middle of typing a word (for ex) and has dismissed completion, this will cause it to re-open on the next keystroke. Based on the previous bug on `{` I think we need to be really careful about where we pop completion open automatically, as if the next character pressed is one that commits completion (enter/tab/etc) then it'll cause random text to appear.

Reply:
I understand your concern, I was actually hesitant to introduce this for the same reasons.

After spending some time testing with and without full alphabet in trigger characters I found that:

> My concern is that if a user is in the middle of typing a word (for ex) and has dismissed completion, this will cause it to re-open on the next keystroke.

This is actually how it works today in 2.7.3, according to my testing. Feel free to test it yourself though.

Adding full alphabet did not change this behavior for me at all. Including if I'm inside strings other than in import (and export) statements.

I suspect if Analyzer doesn't have any suggestions (which is the case inside strings) there is nothing to show, so we don't see it.

> idea of adding the whole alphabet - what was this to try and do?

It is what you said - invoke completion popup when user types a character in `a-z` range. This solves the imports (and exports) issue including when user tries to modify it. And, as far as I can tell, this doesn't change anything else comparing to how it works today.

I will have to spend more time with this to get more confidence, but this is my observation so far.

I'm not sure if it's the best approach in this case, but I thought I'd create this PR as a suggestion to get more opinions from people who know more than me.

Reply:
> This is actually how it works today in 2.7.3, according to my testing. Feel free to test it yourself though.

You're right! However, there is a slight difference. What's happening today is that when you press the next character, Code is re-displaying the previous results but filtered (client-side) and not actually going to the completion provider. With your change, it always goes to the completion provider which means the changes results in a lot more over-the-wire communication (and I guess possibly increased CPU in the analyzer).

I think a better fix would be to trigger on the quotes and review the line to see if it's an import, and then use the mechanism @mjbvz suggested (thanks!) to trigger it for completions. I think this will provide the behaviour we want, without sending a request to the analyzer for a new completion list on every key press.

WDYT? I should have time tomorrow evening to play around with it.

Reply:
> which means the changes results in a lot more over-the-wire communication

Yeah, this is bad.
I think what @mjbvz suggested is a nice improvement and we should definitely do that.

I will update this PR to handle quotes as originally suggested.
Tried to find where "suggest a command" fix can be applied, but didn't find anything at the first look.

Reply:
> Tried to find where "suggest a command" fix can be applied, but didn't find anything at the first look.

I assume you mean where to force completion when the `import "";` snippet is entered? This actually depends a little on whether it's coming from the analysis server or a snippet in Dart Code (I can't remember - however, there's a branch that converts snippets to a completion provider which I want to merge soon...).

For the completion provider, we need to add the command when we build the completion item:

https://github.com/pulyaevskiy/Dart-Code/blob/e0a25bff2dd77faafdbdb5a33b2715f23af460c4/src/providers/dart_completion_item_provider.ts#L119-L134

We'd probably want to build a whitelist so we can choose exactly which ones trigger the completion (so in the first instance, just the one with `import "";` or w/e as the completion text.